presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaPageSinkProvider.java

```diff
@@ -0,0 +1,111 @@
+ */
+package io.prestosql.plugin.kafka;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.prestosql.plugin.kafka.encoder.DispatchingRowEncoderFactory;
+import io.prestosql.plugin.kafka.encoder.EncoderColumnHandle;
+import io.prestosql.plugin.kafka.encoder.RowEncoder;
+import io.prestosql.spi.PrestoException;
+import io.prestosql.spi.connector.ConnectorInsertTableHandle;
+import io.prestosql.spi.connector.ConnectorOutputTableHandle;
+import io.prestosql.spi.connector.ConnectorPageSink;
+import io.prestosql.spi.connector.ConnectorPageSinkProvider;
+import io.prestosql.spi.connector.ConnectorSession;
+import io.prestosql.spi.connector.ConnectorTransactionHandle;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Optional;
+
+import static io.prestosql.plugin.kafka.KafkaErrorCode.KAFKA_AVRO_SCHEMA_ERROR;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class KafkaPageSinkProvider
+        implements ConnectorPageSinkProvider
+{
+    private final DispatchingRowEncoderFactory encoderFactory;
+    private final PlainTextKafkaProducerFactory producerFactory;
+
+    @Inject
+    public KafkaPageSinkProvider(DispatchingRowEncoderFactory encoderFactory, PlainTextKafkaProducerFactory producerFactory)
+    {
+        this.encoderFactory = requireNonNull(encoderFactory, "encoderFactory is null");
+        this.producerFactory = requireNonNull(producerFactory, "producerFactory is null");
+    }
+
+    @Override
+    public ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorOutputTableHandle tableHandle)
+    {
+        throw new UnsupportedOperationException("Table creation is not supported by the kafka connector");
+    }
+
+    @Override
+    public ConnectorPageSink createPageSink(ConnectorTransactionHandle transactionHandle, ConnectorSession session, ConnectorInsertTableHandle tableHandle)
+    {
+        requireNonNull(tableHandle, "tableHandle is null");
+        KafkaTableHandle handle = (KafkaTableHandle) tableHandle;
+
+        ImmutableSet.Builder<EncoderColumnHandle> keyColumns = ImmutableSet.builder();
+        ImmutableSet.Builder<EncoderColumnHandle> messageColumns = ImmutableSet.builder();
+        handle.getColumns().forEach(col -> {
+            if (col.isInternal()) {
+                throw new IllegalArgumentException(format("unexpected internal column handle '%s'", col.getName()));
+            }
+            if (col.isKeyDecoder()) {
+                keyColumns.add(col);
+            }
+            else {
+                messageColumns.add(col);
+            }
+        });
+
+        RowEncoder keyEncoder = encoderFactory.create(
+                handle.getKeyDataFormat(),
+                ImmutableMap.of("dataSchema", getDataSchema(handle.getKeyDataSchemaLocation())),
```

Comment:
Maybe we should have a constant for this?

Reply:
if the encoder params are used just for the avro data schema, would it make sense to just pass an `Optional<String>` that stores the dataSchema? The map seems like overkill, and is kinda clunky, for just one variable.